File: backend/routes/credit.py
from flask import Blueprint, request, jsonify
from backend.models.database import get_db

credit_bp = Blueprint('credit', __name__)

@credit_bp.route('/accounts', methods=['GET'])
def get_accounts():
    db = get_db()
    accounts = db.execute('SELECT * FROM credit_accounts ORDER BY name').fetchall()
    result = []
    for a in accounts:
        a = dict(a)
        txns = db.execute(
            'SELECT type, SUM(amount) as total FROM credit_transactions WHERE account_id = ? GROUP BY type',
            (a['id'],)
        ).fetchall()
        charges = sum(t['total'] for t in txns if t['type'] == 'charge')
        payments = sum(t['total'] for t in txns if t['type'] == 'payment')
        a['computed_balance'] = round(charges - payments, 2)
        a['utilization_pct'] = round((a['computed_balance'] / a['credit_limit'] * 100), 1) if a['credit_limit'] else None
        result.append(a)
    db.close()
    return jsonify(result)

@credit_bp.route('/accounts', methods=['POST'])
def create_account():
    data = request.get_json()
    if not data.get('name'):
        return jsonify({'error': 'name required'}), 400
    db = get_db()
    c = db.cursor()
    c.execute(
        'INSERT INTO credit_accounts (name, last_four, credit_limit) VALUES (?, ?, ?)',
        (data['name'], data.get('last_four', ''), data.get('credit_limit'))
    )
    new_id = c.lastrowid
    db.commit()
    account = db.execute('SELECT * FROM credit_accounts WHERE id = ?', (new_id,)).fetchone()
    db.close()
    return jsonify(dict(account)), 201

@credit_bp.route('/accounts/<int:acct_id>', methods=['DELETE'])
def delete_account(acct_id):
    db = get_db()
    db.execute('DELETE FROM credit_transactions WHERE account_id = ?', (acct_id,))
    db.execute('DELETE FROM credit_accounts WHERE id = ?', (acct_id,))
    db.commit()
    db.close()
    return jsonify({'deleted': acct_id})

@credit_bp.route('/transactions', methods=['GET'])
def get_transactions():
    account_id = request.args.get('account_id')
    db = get_db()
    if account_id:
        rows = db.execute(
            'SELECT ct.*, ca.name as account_name FROM credit_transactions ct JOIN credit_accounts ca ON ct.account_id = ca.id WHERE ct.account_id = ? ORDER BY ct.date DESC, ct.id DESC',
            (account_id,)
        ).fetchall()
    else:
        rows = db.execute(
            'SELECT ct.*, ca.name as account_name FROM credit_transactions ct JOIN credit_accounts ca ON ct.account_id = ca.id ORDER BY ct.date DESC, ct.id DESC'
        ).fetchall()
    db.close()
    return jsonify([dict(r) for r in rows])

@credit_bp.route('/transactions', methods=['POST'])
def create_transaction():
    data = request.get_json()
    required = ['account_id', 'type', 'amount', 'category', 'date']
    for f in required:
        if f not in data:
            return jsonify({'error': f'Missing: {f}'}), 400
    if data['type'] not in ('charge', 'payment'):
        return jsonify({'error': 'type must be charge or payment'}), 400
    db = get_db()
    c = db.cursor()
    c.execute(
        'INSERT INTO credit_transactions (account_id, type, amount, category, description, date) VALUES (?, ?, ?, ?, ?, ?)',
        (data['account_id'], data['type'], abs(float(data['amount'])),
         data['category'], data.get('description', ''), data['date'])
    )
    new_id = c.lastrowid
    db.commit()
    row = db.execute(
        'SELECT ct.*, ca.name as account_name FROM credit_transactions ct JOIN credit_accounts ca ON ct.account_id = ca.id WHERE ct.id = ?',
        (new_id,)
    ).fetchone()
    db.close()
    return jsonify(dict(row)), 201

@credit_bp.route('/transactions/<int:txn_id>', methods=['DELETE'])
def delete_transaction(txn_id):
    db = get_db()
    db.execute('DELETE FROM credit_transactions WHERE id = ?', (txn_id,))
    db.commit()
    db.close()
    return jsonify({'deleted': txn_id})

@credit_bp.route('/stats', methods=['GET'])
def get_stats():
    db = get_db()
    accounts = db.execute('SELECT * FROM credit_accounts').fetchall()
    total_balance = 0
    total_limit = 0
    result = []
    for a in accounts:
        a = dict(a)
        txns = db.execute(
            'SELECT type, SUM(amount) as total FROM credit_transactions WHERE account_id = ? GROUP BY type',
            (a['id'],)
        ).fetchall()
        charges = sum(t['total'] for t in txns if t['type'] == 'charge')
        payments = sum(t['total'] for t in txns if t['type'] == 'payment')
        bal = round(charges - payments, 2)
        a['computed_balance'] = bal
        total_balance += bal
        if a['credit_limit']:
            total_limit += a['credit_limit']
        result.append(a)
    db.close()
    return jsonify({
        'total_balance': round(total_balance, 2),
        'total_limit': round(total_limit, 2),
        'overall_utilization': round((total_balance / total_limit * 100), 1) if total_limit else None,
        'accounts': result
    })

File: frontend/js/modules/credit.js
// modules/credit.js

let creditAccounts = [];
let creditTxns = [];
let selectedAccountId = null;

const CREDIT_CATEGORIES = [
  'Food & Dining', 'Shopping', 'Gas', 'Travel', 'Entertainment',
  'Healthcare', 'Utilities', 'Subscriptions', 'Trading', 'Other'
];

async function renderCredit() {
  const el = document.getElementById('page-credit');
  el.innerHTML = loadingHtml('Loading credit');
  try {
    const [accounts, stats, txns] = await Promise.all([
      api.get('/credit/accounts'),
      api.get('/credit/stats'),
      api.get('/credit/transactions')
    ]);
    creditAccounts = accounts;
    creditTxns = txns;
    renderCreditView(el, accounts, stats, txns);
  } catch (e) {
    el.innerHTML = `<p style="color:var(--red);padding:20px">Error: ${e.message}</p>`;
  }
}

function renderCreditView(el, accounts, stats, txns) {
  const accountCards = accounts.length === 0
    ? `<div style="color:var(--text-3);font-size:0.8rem;padding:20px 0">No credit accounts yet — add one to get started.</div>`
    : accounts.map(a => {
        const utilColor = !a.utilization_pct ? 'var(--text-3)'
          : a.utilization_pct > 80 ? 'var(--red)'
          : a.utilization_pct > 50 ? 'var(--yellow)'
          : 'var(--green)';
        const isSelected = selectedAccountId === a.id;
        return `
          <div class="credit-card ${isSelected ? 'selected' : ''}" onclick="selectAccount(${a.id})">
            <div style="display:flex;justify-content:space-between;align-items:flex-start">
              <div>
                <div style="font-family:var(--font-display);font-weight:700;font-size:1rem">${a.name}</div>
                ${a.last_four ? `<div style="font-size:0.7rem;color:var(--text-3);letter-spacing:0.1em">•••• ${a.last_four}</div>` : ''}
              </div>
              <button class="btn btn-sm btn-danger btn-icon" onclick="event.stopPropagation();deleteAccount(${a.id})">✕</button>
            </div>
            <div style="margin-top:14px">
              <div style="font-family:var(--font-display);font-size:1.6rem;font-weight:800;color:${a.computed_balance > 0 ? 'var(--red)' : 'var(--green)'}">
                $${Math.abs(a.computed_balance).toFixed(2)}
              </div>
              <div style="font-size:0.7rem;color:var(--text-3);margin-top:2px">current balance</div>
            </div>
            ${a.credit_limit ? `
            <div style="margin-top:10px">
              <div style="display:flex;justify-content:space-between;font-size:0.7rem;color:var(--text-3);margin-bottom:4px">
                <span>Utilization</span><span style="color:${utilColor}">${a.utilization_pct}%</span>
              </div>
              <div style="height:4px;background:var(--bg-4);border-radius:2px;overflow:hidden">
                <div style="height:100%;width:${Math.min(a.utilization_pct, 100)}%;background:${utilColor};border-radius:2px;transition:width 0.4s"></div>
              </div>
              <div style="font-size:0.68rem;color:var(--text-3);margin-top:4px">Limit: $${a.credit_limit.toLocaleString()}</div>
            </div>` : ''}
          </div>
        `;
      }).join('');

  const filteredTxns = selectedAccountId
    ? txns.filter(t => t.account_id === selectedAccountId)
    : txns;

  const tableRows = filteredTxns.length === 0
    ? `<tr class="empty-row"><td colspan="6">${emptyStateHtml('▣', 'No transactions yet')}</td></tr>`
    : filteredTxns.map(t => `
        <tr>
          <td>${fmtDate(t.date)}</td>
          <td><span style="font-size:0.75rem;color:var(--text-2)">${t.account_name}</span></td>
          <td><span class="badge ${t.type === 'charge' ? 'badge-red' : 'badge-green'}">${t.type}</span></td>
          <td>${t.category}</td>
          <td>${t.description || '<span class="zero">—</span>'}</td>
          <td class="${t.type === 'charge' ? 'neg' : 'pos'}">${t.type === 'charge' ? '-' : '+'}$${parseFloat(t.amount).toFixed(2)}</td>
          <td>
            <button class="btn btn-sm btn-danger btn-icon" onclick="deleteCreditTxn(${t.id})">✕</button>
          </td>
        </tr>
      `).join('');

  el.innerHTML = `
    <div class="page-header">
      <div>
        <div class="page-title">Credit</div>
        <div class="page-subtitle">Cards & balances</div>
      </div>
      <div style="display:flex;gap:8px;flex-wrap:wrap">
        <button class="btn btn-ghost" onclick="openNewAccountModal()">+ Add Card</button>
        <button class="btn btn-primary" onclick="openNewCreditTxnModal()">+ Log Charge</button>
      </div>
    </div>

    ${stats.total_limit ? `
    <div class="stat-grid" style="margin-bottom:20px">
      <div class="stat-card red">
        <div class="stat-label">Total Owed</div>
        <div class="stat-value neg">$${stats.total_balance.toFixed(2)}</div>
        <div class="stat-sub">across ${accounts.length} card${accounts.length !== 1 ? 's' : ''}</div>
      </div>
      <div class="stat-card">
        <div class="stat-label">Total Limit</div>
        <div class="stat-value">$${stats.total_limit.toLocaleString()}</div>
      </div>
      <div class="stat-card ${(stats.overall_utilization || 0) > 50 ? 'red' : ''}">
        <div class="stat-label">Overall Utilization</div>
        <div class="stat-value ${(stats.overall_utilization || 0) > 50 ? 'neg' : 'pos'}">${stats.overall_utilization ?? '—'}%</div>
        <div class="stat-sub">${(stats.overall_utilization || 0) > 30 ? '⚠ Consider paying down' : '✓ Healthy'}</div>
      </div>
    </div>` : ''}

    <div style="display:grid;grid-template-columns:repeat(auto-fill,minmax(240px,1fr));gap:14px;margin-bottom:24px">
      ${accountCards}
    </div>

    <div class="action-row">
      <div style="font-size:0.75rem;color:var(--text-2)">
        ${selectedAccountId
          ? `Showing: <strong>${accounts.find(a => a.id === selectedAccountId)?.name}</strong> <button class="btn btn-sm btn-ghost" onclick="selectAccount(null)" style="margin-left:8px">Clear filter</button>`
          : 'All transactions'}
      </div>
      <span style="font-size:0.72rem;color:var(--text-3)">${filteredTxns.length} transaction${filteredTxns.length !== 1 ? 's' : ''}</span>
    </div>

    <div class="table-wrap">
      <table>
        <thead>
          <tr>
            <th>Date</th>
            <th>Card</th>
            <th>Type</th>
            <th>Category</th>
            <th>Description</th>
            <th>Amount</th>
            <th></th>
          </tr>
        </thead>
        <tbody>${tableRows}</tbody>
      </table>
    </div>
  `;
}

function selectAccount(id) {
  selectedAccountId = id;
  renderCredit();
}

function openNewAccountModal() {
  openModal(`
    <div class="modal-title">Add Credit Card</div>
    <div class="form-grid">
      <div class="form-group">
        <label class="form-label">Card Name</label>
        <input id="ca-name" class="form-input" placeholder="Chase Sapphire, Discover…">
      </div>
      <div class="form-group">
        <label class="form-label">Last 4 Digits (optional)</label>
        <input id="ca-four" class="form-input" maxlength="4" placeholder="1234">
      </div>
      <div class="form-group">
        <label class="form-label">Credit Limit ($, optional)</label>
        <input id="ca-limit" class="form-input" type="number" step="any" placeholder="500">
      </div>
    </div>
    <button class="btn btn-primary" onclick="submitNewAccount()" style="width:100%">Add Card</button>
  `);
}

async function submitNewAccount() {
  const name = document.getElementById('ca-name').value.trim();
  if (!name) { showToast('Card name required', 'error'); return; }
  try {
    await api.post('/credit/accounts', {
      name,
      last_four: document.getElementById('ca-four').value,
      credit_limit: document.getElementById('ca-limit').value || null
    });
    closeModal();
    showToast('Card added ✓', 'success');
    renderCredit();
  } catch (e) {
    showToast(e.message, 'error');
  }
}

async function deleteAccount(id) {
  confirmAction('Delete this card and ALL its transactions?', async () => {
    try {
      await api.del(`/credit/accounts/${id}`);
      if (selectedAccountId === id) selectedAccountId = null;
      showToast('Card deleted', 'success');
      renderCredit();
    } catch (e) {
      showToast(e.message, 'error');
    }
  });
}

function openNewCreditTxnModal() {
  if (creditAccounts.length === 0) {
    showToast('Add a credit card first', 'error');
    openNewAccountModal();
    return;
  }
  window._creditTxnType = 'charge';
  openModal(`
    <div class="modal-title">Log Credit Transaction</div>
    <div class="form-group" style="margin-bottom:14px">
      <label class="form-label">Type</label>
      <div class="radio-group">
        <div id="ct-charge" class="radio-btn active-expense" onclick="setCreditTxnType('charge')">Charge</div>
        <div id="ct-payment" class="radio-btn" onclick="setCreditTxnType('payment')">Payment</div>
      </div>
    </div>
    <div class="form-grid">
      <div class="form-group">
        <label class="form-label">Card</label>
        <select id="ct-acct" class="form-select">
          ${creditAccounts.map(a => `<option value="${a.id}">${a.name}${a.last_four ? ' ••'+a.last_four : ''}</option>`).join('')}
        </select>
      </div>
      <div class="form-group">
        <label class="form-label">Amount ($)</label>
        <input id="ct-amount" class="form-input" type="number" step="any" placeholder="0.00">
      </div>
      <div class="form-group">
        <label class="form-label">Date</label>
        <input id="ct-date" class="form-input" type="date" value="${todayISO()}">
      </div>
      <div class="form-group">
        <label class="form-label">Category</label>
        <select id="ct-cat" class="form-select">
          ${CREDIT_CATEGORIES.map(c => `<option>${c}</option>`).join('')}
        </select>
      </div>
    </div>
    <div class="form-group" style="margin-bottom:18px">
      <label class="form-label">Description (optional)</label>
      <input id="ct-desc" class="form-input" placeholder="Merchant, notes…">
    </div>
    <button class="btn btn-primary" onclick="submitNewCreditTxn()" style="width:100%">Log Transaction</button>
  `);
}

function setCreditTxnType(type) {
  window._creditTxnType = type;
  document.getElementById('ct-charge').className = `radio-btn ${type === 'charge' ? 'active-expense' : ''}`;
  document.getElementById('ct-payment').className = `radio-btn ${type === 'payment' ? 'active-income' : ''}`;
}

async function submitNewCreditTxn() {
  const amount = document.getElementById('ct-amount').value;
  const date = document.getElementById('ct-date').value;
  if (!amount || !date) { showToast('Fill in required fields', 'error'); return; }
  try {
    await api.post('/credit/transactions', {
      account_id: parseInt(document.getElementById('ct-acct').value),
      type: window._creditTxnType,
      amount, date,
      category: document.getElementById('ct-cat').value,
      description: document.getElementById('ct-desc').value
    });
    closeModal();
    showToast('Transaction logged ✓', 'success');
    renderCredit();
  } catch (e) {
    showToast(e.message, 'error');
  }
}

async function deleteCreditTxn(id) {
  confirmAction('Delete this transaction?', async () => {
    try {
      await api.del(`/credit/transactions/${id}`);
      showToast('Deleted', 'success');
      renderCredit();
    } catch (e) {
      showToast(e.message, 'error');
    }
  });
}